From adfa3df2268edefd74efe293e5c3b4b09b5d38ec Mon Sep 17 00:00:00 2001
From: Objectstor Dev <objectstor@gmail.com>
Date: Wed, 4 Jan 2023 13:11:12 +0000
Subject: [PATCH 2/4] SkyBroker AWS S3 patch

---
 providers/s3/s3.go | 23 +++++++++++++++--------
 1 file changed, 15 insertions(+), 8 deletions(-)

diff --git a/providers/s3/s3.go b/providers/s3/s3.go
index 0a7369c..0e94ec8 100644
--- a/providers/s3/s3.go
+++ b/providers/s3/s3.go
@@ -26,8 +26,10 @@ import (
 	"github.com/pkg/errors"
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/common/version"
+
 	"gitlab.gs.com/sre/g/crypto/krb"
 	"gitlab.gs.com/sre/g/net/aws/miniocreds"
+	"gitlab.gs.com/sre/g/net/aws/skybrokercreds"
 	ghttp "gitlab.gs.com/sre/g/net/http"
 	"gitlab.gs.com/sre/g/process/refresh"
 
@@ -150,13 +152,13 @@ type Config struct {
 }
 
 type GsSkyConfig struct {
-	Keytab                 string          `yaml:"keytab"`
-	Principal              string          `yaml:"principal"`
-	RoleARN                string          `yaml:"role_arn"`
-	GsssoTokenLifetime     model.Duration  `yaml:"gssso_token_lifetime"`
-	GsssoRefresh           GsRefreshConfig `yaml:"gssso_refresh"`
-	SkyBrokerRefresh       GsRefreshConfig `yaml:"isky_broker_refresh"`
-	SkyBrokerTokenWaitTime model.Duration  `yaml:"sky_broker_token_wait_timeout"`
+	Keytab                 string                 `yaml:"keytab"`
+	Principal              string                 `yaml:"principal"`
+	RoleARN                string                 `yaml:"role_arn"`
+	GsssoTokenLifetime     model.Duration         `yaml:"gssso_token_lifetime"`
+	GsssoRefresh           GsRefreshConfig        `yaml:"gssso_refresh"`
+	SkyBrokerRefresh       SkyBrokerRefreshConfig `yaml:"isky_broker_refresh"`
+	SkyBrokerTokenWaitTime model.Duration         `yaml:"sky_broker_token_wait_timeout"`
 }
 
 type GsRefreshConfig struct {
@@ -168,6 +170,11 @@ type GsRefreshConfig struct {
 	TokenExpiryGap      model.Duration `yaml:"token_expiry_gap"`
 }
 
+type SkyBrokerRefreshConfig struct {
+	GsRefreshConfig          `yaml:",inline"`
+	SkyBrokerDesiredLifetime model.Duration `yaml:"sts_desired_lifetime"`
+}
+
 // SSEConfig deals with the configuration of SSE for Minio. The following options are valid:
 // kmsencryptioncontext == https://docs.aws.amazon.com/kms/latest/developerguide/services-s3.html#s3-encryption-context
 type SSEConfig struct {
@@ -273,7 +280,7 @@ func NewBucketWithConfig(logger log.Logger, config Config, component string) (*B
 			RetryMaxGrowSteps:   config.SkyConfig.SkyBrokerRefresh.RetryMaxGrowSteps,
 			FetchTimeout:        time.Duration(config.SkyConfig.SkyBrokerRefresh.FetchTimeout),
 			TokenExpiryGap:      time.Duration(config.SkyConfig.SkyBrokerRefresh.TokenExpiryGap),
-		})
+		}, skybrokercreds.WithDesiredLifetime(time.Duration(config.SkyConfig.SkyBrokerRefresh.SkyBrokerDesiredLifetime)))
 		if err != nil {
 			return nil, errors.Wrap(err, "constructing refreshing credentials provider for SkyBroker S3 authentication")
 		}
-- 
2.37.2

